src/nvim/tui/tui.c

```diff
@@ -0,0 +1,617 @@
+    }
+  } else {
+    while (*str) {
+      if (*str == '<') {
+        len += (size_t)snprintf(buf + len, sizeof(buf) - len, "<lt>");
+      } else {
+        buf[len++] = *str;
+      }
+      str++;
+    }
+  }
+
+  buf[len] = 0;
+  input_enqueue((String) {.data = buf, .size = len});
+}
+
+static void mouse_event(TickitTerm *tt, TickitMouseEventType type,
+    int button, int row, int col, int mod)
+{
+  size_t len = 0;
+  char buf[64];
+  char buf2[64];
+
+  if (button == 1) {
+    len = (size_t)snprintf(buf , sizeof(buf), "Left");
+  } else if (button == 2) {
+    len = (size_t)snprintf(buf, sizeof(buf), "Middle");
+  } else if (button == 3) {
+    len = (size_t)snprintf(buf, sizeof(buf), "Right");
+  } else {
+    return;
+  }
+
+  if (type == TICKIT_MOUSEEV_PRESS) {
+    snprintf(buf + len, sizeof(buf) - len, "Mouse");
+  } else if (type == TICKIT_MOUSEEV_DRAG) {
+    snprintf(buf + len, sizeof(buf) - len, "Drag");
+  } else if (type == TICKIT_MOUSEEV_WHEEL) {
+    if (button == TICKIT_MOUSEWHEEL_UP) {
+      snprintf(buf, sizeof(buf), "ScrollWheelUp");
+    } else {
+      snprintf(buf, sizeof(buf), "ScrollWheelDown");
+    }
+  }
+
+  len = translate_key(buf2, sizeof(buf2), buf, mod);
+  len += (size_t)snprintf(buf2 + len, sizeof(buf2) - len, "<%d,%d>", col, row);
+  input_enqueue((String) {.data = buf2, .size = len});
+}
+
+static void event(TickitTerm *tt, TickitEventType ev, TickitEvent *args,
+    void *data)
+{
+  switch(ev) {
+  case TICKIT_EV_RESIZE:
+  case TICKIT_EV_CHANGE:
+  case TICKIT_EV_UNBIND:
+    break;
+
+  case TICKIT_EV_KEY:
+    key_event(tt, (TickitKeyEventType)args->type, args->str, args->mod);
+    break;
+
+  case TICKIT_EV_MOUSE:
+    mouse_event(tt, (TickitMouseEventType)args->type, args->button,
+        args->line, args->col, args->mod);
+    break;
+  }
+}
+
+static void try_resize(Event ev)
+{
+  UI *ui = ev.data;
+  TUIData *data = ui->data;
+  data->old_height = ui->height;
+  tickit_term_refresh_size(data->tt);
+  tickit_term_get_size(data->tt, &ui->height, &ui->width);
+  ui_refresh();
+}
+
+static void sigwinch_cb(uv_signal_t *handle, int signum)
+{
+  // Queue the event because resizing can result in recursive event_poll calls
+  event_push((Event) {
+    .data = handle->data,
+    .handler = try_resize
+  }, false);
+}
+
+static void timer_cb(uv_timer_t *handle)
+{
+  UI *ui = handle->data;
+  TUIData *data = ui->data;
+  tickit_term_input_check_timeout(data->tt);
+}
+
+static void poll_cb(uv_poll_t *handle, int status, int events)
+{
+  if (status < 0) {
+    input_done();
+    return;
+  }
+
+  UI *ui = handle->data;
+  TUIData *data = ui->data;
+  tickit_term_input_readable(data->tt);
+  int timeout = tickit_term_input_check_timeout(data->tt);
+
+  if (timeout >= 0) {
+    uv_timer_start(&data->timer_handle, timer_cb, (uint32_t)timeout, 0);
+  }
+}
+
+static bool attrs_differ(HlAttrs a1, HlAttrs a2)
+{
+  return a1.foreground != a2.foreground || a1.background != a2.background
```

Comment:
Answer: no. :-(  It's a real tempting optimization to make, but it doesn't work unless you can be certain that the padding is zeroized--which is probably not a good assumption to make, even if our current paths all do it.